File: 1377. Frog Position After T Seconds/Solution.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {
    public double frogPosition(int n, int[][] edges, int t, int target) {
        List<Integer>[] tree = new ArrayList[n + 1];
        Arrays.setAll(tree, e -> new ArrayList<>());
        tree[1].add(0);
        for (int[] edge : edges) {
            tree[edge[0]].add(edge[1]);
            tree[edge[1]].add(edge[0]);
        }
        double choices = dfs(tree, 0, 1, target, t);
        return choices != 0 ? 1.0 / choices : 0;
    }

    private int dfs(List<Integer>[] tree, int parent, int vertice, int target, int t) {
        if (t == 0) {
            return vertice == target ? 1 : 0;
        }
        if (vertice == target) {
            return tree[vertice].size() == 1 ? 1 : 0;
        }
        for (Integer neighbour : tree[vertice]) {
            if (neighbour != parent) {
                int choices = dfs(tree, vertice, neighbour, target, t - 1);
                if (choices != 0) {
                    return choices * (tree[vertice].size() - 1);
                }
            }
        }
        return 0;
    }
}

File: 1377. Frog Position After T Seconds/Solution.py
from typing import List


class Solution:
    def frogPosition(self, n: int, edges: List[List[int]], t: int, target: int) -> float:
        tree = [[] for _ in range(n + 1)]
        tree[1] = [0]
        for x, y in edges:
            tree[x].append(y)
            tree[y].append(x)

        def dfs(vertice: int, parent: int, left_t: int) -> int:
            if left_t == 0:
                return vertice == target
            if vertice == target:
                return len(tree[vertice]) == 1
            for neighbour in tree[vertice]:
                if neighbour == parent:
                    continue
                choices = dfs(neighbour, vertice, left_t - 1)
                if choices:
                    return choices * (len(tree[vertice]) - 1)
            return 0

        choices = dfs(1, 0, t)
        return 1 / choices if choices else 0
